google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/RateLimitingServerStreamingCallable.java

```diff
@@ -0,0 +1,149 @@
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import org.threeten.bp.Duration;
+
+class RateLimitingServerStreamingCallable
+    extends ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> {
+  private static final Logger logger =
+      Logger.getLogger(RateLimitingServerStreamingCallable.class.getName());
+  private static final long DEFAULT_QPS = 10;
+  // Default interval before changing the QPS on error response
+  private static Duration DEFAULT_PERIOD = Duration.ofSeconds(10);
+  private final RateLimiter limiter;
+  private final AtomicLong lastQpsChangeTime = new AtomicLong(System.currentTimeMillis());
+  private final ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> innerCallable;
+  private static final double MIN_QPS = 0.1;
+  private static final double MAX_QPS = 100_000;
+  @VisibleForTesting static final double MIN_FACTOR = 0.7;
+  private static final double MAX_FACTOR = 1.3;
+
+  RateLimitingServerStreamingCallable(
+      @Nonnull ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> innerCallable) {
+    this.limiter = RateLimiter.create(DEFAULT_QPS);
+    this.innerCallable = Preconditions.checkNotNull(innerCallable, "Inner callable must be set");
+    logger.info("Rate limiting is enabled with initial QPS of " + limiter.getRate());
+  }
+
+  @Override
+  public void call(
+      MutateRowsRequest request,
+      ResponseObserver<MutateRowsResponse> responseObserver,
+      ApiCallContext context) {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    limiter.acquire();
+    stopwatch.stop();
+    if (context.getTracer() instanceof BigtableTracer) {
+      ((BigtableTracer) context.getTracer())
+          .batchRequestThrottled(stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    }
+    RateLimitingResponseObserver innerObserver = new RateLimitingResponseObserver(responseObserver);
+    innerCallable.call(request, innerObserver, context);
+  }
+
+  class RateLimitingResponseObserver extends SafeResponseObserver<MutateRowsResponse> {
+    private final ResponseObserver<MutateRowsResponse> outerObserver;
+
+    RateLimitingResponseObserver(ResponseObserver<MutateRowsResponse> observer) {
+      super(observer);
+      this.outerObserver = observer;
+    }
+
+    @Override
+    protected void onStartImpl(final StreamController controller) {
+      outerObserver.onStart(controller);
+    }
+
+    @Override
+    protected void onResponseImpl(MutateRowsResponse response) {
+      if (response.hasRateLimitInfo()) {
+        RateLimitInfo info = response.getRateLimitInfo();
+        // RateLimitInfo is an optional field. However, proto3 sub-message field always
+        // have presence even thought it's marked as "optional". Check the factor and
+        // period to make sure they're not 0.
+        if (info.getFactor() != 0 && info.getPeriod().getSeconds() != 0) {
+          updateQps(info.getFactor(), Duration.ofSeconds(info.getPeriod().getSeconds()));
```

Comment:
I think it would be safer to use `com.google.protobuf.util.Durations.toSeconds()`